vignettes/invariants.Rmd

```diff
@@ -475,36 +470,55 @@ with_df(df[[1]] <- NULL)
 with_df2(df2[[2]] <- NULL)
 ```
 
-### Definition of `x$name <- a`.
+### Definition of `x$name <- a`
 
 `x$name <- a` and `x$"name" <- a` are equivalent to `x[["name"]] <- a`.[^column-assign-symmetry]
 
-[^column-assign-symmetry]: `$` behaves almost completely symmetrical to `[[` when comparing subsetting and subassignment.
+[^column-assign-symmetry]: `$` behaves almost completely symmetrically to `[[` when comparing subsetting and subassignment.
 
 ```{r dollar-equivalent-to-subset-assign, dftbl = TRUE}
 with_df(df$n <- 0)
-with_df(df$"n" <- 0)
 with_df(df[["n"]] <- 0)
 ```
 
-`$` does not implement partial matching but creates a new column.
+`$<-` does not perform partial matching.
 
 ```{r dollar-equivalent-to-subset-assign-pmatch, dftbl = TRUE}
 with_df(df$l <- 0)
 with_df(df[["l"]] <- 0)
 ```
 
-## Column subassignment
+## Column subassignment: `x[j] <- a`
+
+* If `j` is missing, it's replaced with `seq_along(x)`
+* If `j` is logical, it's converted to numeric with `seq_along(x)[j]`.
 
-### Definition of `x[j] <- list(...)`.
+### `a` is a list or data frame
 
-First, if `j` is not a character vector, it is converted to numeric indices. Then, `x[j] <- a` is equivalent to performing `x[[j_1]] <- a[[1]]`, `x[[j_2]] <- a[[2]]` etc.
-When the same column is updated/removed several times, the last update wins.
-If `a` contains `NULL` values, the corresponding columns are removed *after* updating.
+If `inherits(a, "list")` or `inherits(a, "data.frame")` is `TRUE`, then `x[j] <- a` is equivalent to `x[[j[[1]]] <- a[[1]]`, `x[[j[[2]]]] <- a[[2]]`, ...
 
 ```{r bracket-assign-def, dftbl = TRUE}
 with_df(df[1:2] <- list("x", 4:1))
 with_df(df[c("li", "x", "c")] <- list("x", 4:1, NULL))
+```
+
+If `length(a)` equals 1, then it is recycled to the same length as `j`.
+
+```{r bracket-assign-recycles, dftbl = TRUE}
+with_df(df[1:2] <- list(1))
+with_df(df[1:2] <- list(0, 0, 0))
+with_df(df[1:3] <- list(0, 0))
+```
+
+If the same column is updated/removed several times, the last update wins.
```

Comment:
I wasn't aware of this:

```suggestion
An attempt to update the same column twice gives an error.
```

``` r
df <- data.frame(a = 1:3)
df[c(1, 1)] <- 2
#> Error in `[<-.data.frame`(`*tmp*`, c(1, 1), value = 2): duplicate subscripts for columns
```

<sup>Created on 2019-10-01 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>